### PR TITLE
Add gitlab-runner

### PR DESCRIPTION
**Pull request by duncanmmacleod**
This adds a recipe for gitlab-runner.

Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your
      recipe (see
      [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos)
      for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

File: recipes/gitlab-runner/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+{% set name = "gitlab-runner" %}
+{% set version = "12.1.0" %}
+
+package:
+  name: "{{ name|lower }}"
+  version: "{{ version }}"
+
+source:
+  git_url: "https://gitlab.com/gitlab-org/gitlab-runner.git"
```
Comment (xhochy) [inline recipes/gitlab-runner/meta.yaml]: Can you use a tarball here?

Comment (marcelotrevisani) [inline recipes/gitlab-runner/meta.yaml]: You can replace it by
```suggestion
  git_url: "https://gitlab.com/gitlab-org/gitlab-runner/-/archive/v{{ version }}/gitlab-runner-v{{ version }}.tar.gz"
```

Comment (duncanmmacleod) [reply recipes/gitlab-runner/meta.yaml]: @marcelotrevisani, @xhochy, I am following the build procedure that is used for the semi-official [homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/gitlab-runner.rb), which includes inserting the git hash into the build output. I have no idea whether this is a requirement of the build, or just a useful piece of metadata when you are building from git.

This is why I have not used the tarball, but I am happy to receive advice on how to proceed.

**Comment by duncanmmacleod** [comment]
The appveyor build failed (stopped) and the equivalent azure windows build is hanging. Can someone suggest why this might be, and how I can update the recipe to prevent a repeat?

File: recipes/gitlab-runner/meta.yaml (hunk 1 above)
Comment (dbast) [reply recipes/gitlab-runner/meta.yaml]: I guess you can set the revision to whatever you like ... here it is set to conda-forge: https://github.com/conda-forge/fzf-feedstock/blob/master/recipe/meta.yaml 

This recipe also contains some different ideas building a go-recipe.

**Comment by sodre** [comment]
@duncanmmacleod, may I suggest looking at #8919 for an idea on how you can build `gitlab-runner`

**Comment by sodre** [comment]
@duncanmmacleod, have you had a chance to look at the [PR](https://github.com/duncanmmacleod/staged-recipes/pull/2) I left for you? It should make it this recipe relatively simple and easy to merge.

File: recipes/gitlab-runner/meta.yaml (hunk 1 above)
Comment (sodre) [reply recipes/gitlab-runner/meta.yaml]: We are using a tarball now.

Comment (sodre) [reply recipes/gitlab-runner/meta.yaml]: Revision is set to `deadbeef`

**Comment by sodre** [comment]
@conda-forge/staged-recipes the gitlab-runner recipe is ready for review again.

File: recipes/gitlab-runner/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  - folder: {{ pkg_src }}
+    url: https://{{ import_path }}/-/archive/v{{ version }}/{{ name }}-v{{ version }}.tar.bz2
+    sha256: 49e7993e1eabc362ec99562e84931e699b77948efb6669f3e98a5dc40f2b2b66
+
+build:
+  number: 0
+  script:
+    - pushd {{ pkg_src }}
+    - go install -v
+        -ldflags "-X {{import_path}}/common.NAME='{{ name }}'
```
Comment (duncanmmacleod) [inline recipes/gitlab-runner/meta.yaml]:
```suggestion
        -ldflags "-X {{ import_path }}/common.NAME='{{ name }}'
```

File: recipes/gitlab-runner/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                  -X {{import_path}}/common.VERSION='{{ version }}'
```
Comment (duncanmmacleod) [inline recipes/gitlab-runner/meta.yaml]:
```suggestion
                  -X {{ import_path }}/common.VERSION='{{ version }}'
```

File: recipes/gitlab-runner/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                  -X {{import_path}}/common.REVISION='deadbeef'
```
Comment (duncanmmacleod) [inline recipes/gitlab-runner/meta.yaml]:
```suggestion
                  -X {{ import_path }}/common.REVISION='deadbeef' 
```

File: recipes/gitlab-runner/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+build:
+  number: 0
+  script:
+    - pushd {{ pkg_src }}
+    - go install -v
+        -ldflags "-X {{import_path}}/common.NAME='{{ name }}'
+                  -X {{import_path}}/common.VERSION='{{ version }}'
+                  -X {{import_path}}/common.REVISION='deadbeef' 
+                  -X {{import_path}}/common.BRANCH='{{ branch }}'
```
Comment (duncanmmacleod) [inline recipes/gitlab-runner/meta.yaml]:
```suggestion
                  -X {{ import_path }}/common.BRANCH='{{ branch }}'
```

File: recipes/gitlab-runner/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                  -X {{import_path}}/common.BUILT='2019-08-01T00:00:00z'"
```
Comment (duncanmmacleod) [inline recipes/gitlab-runner/meta.yaml]:
```suggestion
                  -X {{ import_path }}/common.BUILT='2019-08-01T00:00:00z'"
```

Also, is there a way to get `conda build` to automatically insert the build date for us?

Comment (sodre) [reply recipes/gitlab-runner/meta.yaml]: Sorry about that. I had not noticed you had the build-date logic for windows and linux in your original commit. I added it back into the recipe.

Comment (sodre) [reply recipes/gitlab-runner/meta.yaml]: The date issue is fixed now as well. 

This recipe is ready for review/merge.

**Comment by sodre** [comment]
@beckermr, would you mind doing a gitlab-runner as well?

@conda-forge/help-go @conda-forge/staged-recipes

**Comment by beckermr** [comment]
GO packaging is well outside my wheelhouse. I'd prefer to have someone else look at this. @conda-forge/staged-recipes can someone who knows about GO please review this?

**Comment by marcelotrevisani** [comment]
Not my area as well...
Maybe @ocefpaf could take a look on that?

**Comment by marcelotrevisani** [comment]
but it looks fine for me

**Comment by dbast** [comment]
The windows build (both Azure and appveyor) do not look good:
```
Unable to parse meta.yaml file

Error Message:
--> while parsing a quoted scalar
-->   in "<unicode string>", line 43, column 17
--> found unknown escape character
-->   in "<unicode string>", line 43, column 21

ERROR D:\a\1\s\.ci_support\compute_build_graph.py:add_recipe_to_graph(221): Exception raised!
Traceback (most recent call last):
  File "C:\Miniconda\lib\site-packages\conda_build\metadata.py", line 222, in yamlize
    loaded_data = yaml.load(data, Loader=loader)
  File "C:\Miniconda\lib\site-packages\yaml\__init__.py", line 114, in load
    return loader.get_single_data()
  File "C:\Miniconda\lib\site-packages\yaml\constructor.py", line 41, in get_single_data
    node = self.get_single_node()
  File "ext/_yaml.pyx", line 707, in _yaml.CParser.get_single_node
  File "ext/_yaml.pyx", line 725, in _yaml.CParser._compose_document
  File "ext/_yaml.pyx", line 776, in _yaml.CParser._compose_node
  File "ext/_yaml.pyx", line 890, in _yaml.CParser._compose_mapping_node
  File "ext/_yaml.pyx", line 776, in _yaml.CParser._compose_node
  File "ext/_yaml.pyx", line 890, in _yaml.CParser._compose_mapping_node
  File "ext/_yaml.pyx", line 732, in _yaml.CParser._compose_node
  File "ext/_yaml.pyx", line 905, in _yaml.CParser._parse_next_event
yaml.scanner.ScannerError: while parsing a quoted scalar
  in "<unicode string>", line 43, column 17
found unknown escape character
  in "<unicode string>", line 43, column 21
```

**Comment by dbast** [comment]
Now it could parse the recipe also during win build. The error is now:
```
The process tried to write to a nonexistent pipe.
Terminate batch job (Y/N)? 
invalid value "-X gitlab.com/gitlab-org/gitlab-runner/common.NAME='gitlab-runner' -X gitlab.com/gitlab-org/gitlab-runner/common.VERSION='12.1.0' -X gitlab.com/gitlab-org/gitlab-runner/common.REVISION='deadbeef' -X gitlab.com/gitlab-org/gitlab-runner/common.BRANCH='12-1-stable' -X gitlab.com/gitlab-org/gitlab-runner/common.BUILT='Enter the new date: (mm-dd-yy) '" for flag -ldflags: unterminated ' string
usage: go install [-i] [build flags] [packages]
Run 'go help install' for details.
```

**Comment by sodre** [comment]
I'll work on it...

**Comment by sodre** [comment]
the problem is the date command on windows....

File: recipes/gitlab-runner/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    sha256: 49e7993e1eabc362ec99562e84931e699b77948efb6669f3e98a5dc40f2b2b66
+
+build:
+  number: 0
+  script:
+    - pushd {{ pkg_src }}
+    #- for /F "delims=" %%i in ('date.exe /t +%Y-%m-%dT%H:%M:%S%z') DO set BUILT="%%i"  # [win]
+    - set BUILT="2019-08-14T00:00:00+0000"  # [win]
+    - export BUILT=$(date -u +%Y-%m-%dT%H:%M:%S%z)  # [unix]
+    {% set flags = dict(VERSION=version, REVISION="deadbeef", BRANCH=branch, BUILD="%BUILT%" if win else "$BUILT") %}
```
Comment (dbast) [inline recipes/gitlab-runner/meta.yaml]: Guessing from build log: the key should be BUILT instead if BUILD.

Comment (sodre) [reply recipes/gitlab-runner/meta.yaml]: I thought so as well, until I checked upstream:

https://gitlab.com/gitlab-org/gitlab-runner/blob/master/Makefile#L51

**Comment by sodre** [comment]
@dbast, Thanks for helping debug the issue. The recipe should be ready for review/merge again.

File: recipes/gitlab-runner/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  - folder: {{ pkg_src }}
+    url: https://{{ import_path }}/-/archive/v{{ version }}/{{ name }}-v{{ version }}.tar.bz2
+    sha256: 49e7993e1eabc362ec99562e84931e699b77948efb6669f3e98a5dc40f2b2b66
+
+build:
+  number: 0
+  script:
+    - pushd {{ pkg_src }}
+    #- for /F "delims=" %%i in ('date.exe /t +%Y-%m-%dT%H:%M:%S%z') DO set BUILT="%%i"  # [win]
+    - set BUILT="2019-08-14T00:00:00+0000"  # [win]
```
Comment (dbast) [inline recipes/gitlab-runner/meta.yaml]: Is that hard coded date for windows intended?

**Comment by sodre** [comment]
Yes, I rather not fight any longer with it. I’ll try a different fix in the main repo.

**Comment by dbast** [comment]
Wow, finally. Took a while to figure out that datetime is available in the jinja context. Sorry for directly committing here.

The built timestamp is now generated without any date mangling code. Also the revision is set to the build number, which together with the version number fully represents the conda build of the package.

Feel free to update/change/remove the commit. Gonna merge it, if you are fine with it.

**Comment by sodre** [comment]
Lgtm! Go for it.

**Pull request merged**